crates/walrus-utils/src/tracing_sampled.rs

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pub use humantime::parse_duration;
+pub use once_cell::sync::Lazy;
+/// Logs the given statement at most once per specified interval for each call site.
+/// Relies on a static variable per invocation location to track the last emission time.
+/// Includes count of skipped events in the log message when an event is finally emitted.
+///
+/// # Arguments
+///
+/// * `$interval`: A `std::time::Duration` specifying the minimum interval between logs.
+/// * `$level:ident`: The tracing level identifier (
+///     e.g., `info`, `debug`, `warn`, `error`, `trace`).
+/// * `$($fields_and_message:tt)*`: The remaining arguments passed directly to the underlying macro.
+///
+/// # Returns
+///
+/// Returns `true` if a log message was emitted by this specific call, `false` otherwise.
+///
+/// # Example
+///
+/// ```rust
+/// use std::time::Duration;
+/// use tracing;
+/// use walrus_utils::tracing_sampled;
+///
+/// fn process_item(item_id: u32) {
+///     // Log processing start max once every 5 seconds for this function call site.
+///     tracing_sampled::info!("5s", item_id, "Starting processing");
+///     // Simulate work
+///     thread::sleep(Duration::from_millis(100));
+///     // Log completion max once every 30 seconds, including skipped count.
+///     tracing_sampled::debug!("30s", item_id, "Finished processing item");
+/// }
+/// ```
+#[macro_export]
+macro_rules! log {
+    ($interval:expr, $level:ident, $($fields_and_message:tt)*) => {
+        {
+            use std::sync::atomic::{AtomicU64, Ordering};
+            use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+            static LAST_LOGGED_MS: $crate::tracing_sampled::Lazy<AtomicU64> =
+                $crate::tracing_sampled::Lazy::new(|| AtomicU64::new(0));
+            static SKIPPED_COUNT: $crate::tracing_sampled::Lazy<AtomicU64> =
+                $crate::tracing_sampled::Lazy::new(|| AtomicU64::new(0));
+
+            let now_ms = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap_or(Duration::from_secs(0))
+                .as_millis() as u64;
+
+            let last_ms = LAST_LOGGED_MS.load(Ordering::Relaxed);
+            let interval_ms =
+                $crate::tracing_sampled::parse_duration($interval).unwrap().as_millis() as u64;
+
+            let should_log = last_ms == 0 || now_ms.saturating_sub(last_ms) >= interval_ms;
+
+            let mut did_log = false;
+
+            if should_log {
+
+                match LAST_LOGGED_MS.compare_exchange(
+                    last_ms,
+                    now_ms,
+                    Ordering::Relaxed,
+                    Ordering::Relaxed
+                ) {
+                    Ok(_) => {
+                        let skipped = SKIPPED_COUNT.swap(0, Ordering::Relaxed);
+
+                        if skipped > 0 {
+                            tracing::$level!(skipped_events = skipped, $($fields_and_message)*);
```

Comment:
```suggestion
                            tracing::$level!(tracing_samples_skipped = skipped, $($fields_and_message)*);
```
Since this log will appear to come from some other module, it maybe be worth making the attribute `skipped_events` more explicitly named, so that it's not confounded with local variables when reading logs.